panels/sharing/cc-sharing-panel.c

```diff
@@ -646,6 +660,154 @@ cc_sharing_panel_setup_media_sharing_dialog (CcSharingPanel *self)
+    }
+  else
+    {
+      g_dbus_proxy_call_sync (self->priv->companion_app_avahi_helper_proxy,
+                              "ExitDiscoverableMode",
+                              NULL,
+                              G_DBUS_CALL_FLAGS_NONE, -1,
+                              NULL, &error);
+    }
+
+  if (error != NULL)
+    {
+      g_critical ("Unable to set the enabled state of content sharing: %s", error->message);
+      g_error_free (error);
+    }
+}
+
+static void
+on_content_sharing_proxy_properties_changed (GDBusProxy *proxy,
+                                             GVariant *changed_properties,
+                                             GStrv invalidated_properties,
+                                             CcSharingPanel *self)
+{
+  GVariant *discoverable_prop;
+  gboolean content_sharing_enabled;
+  GtkWidget *w;
+
+  discoverable_prop = g_variant_lookup_value (changed_properties, "Discoverable", G_VARIANT_TYPE_BOOLEAN);
+  if (!discoverable_prop)
+    return;
+
+  content_sharing_enabled = g_variant_get_boolean (discoverable_prop);
+  g_variant_unref (discoverable_prop);
+
+  w = GTK_WIDGET (gtk_builder_get_object (self->priv->builder, "content-sharing-switch"));
+  gtk_switch_set_active (GTK_SWITCH (w), content_sharing_enabled);
+}
+
+static gboolean
+cc_sharing_panel_check_content_sharing_available (void)
+{
+  GError              *error = NULL;
+  FlatpakInstallation *installation = flatpak_installation_new_system (NULL, &error);
+  FlatpakInstalledRef *ref = NULL;
+  gboolean ret = TRUE; 
+
+  if (installation == NULL)
+    {
+      g_message ("Unexpected error occurred when loading flatpak installation: %s", error->message);
+      g_clear_error (&error);
+      return FALSE;
+    }
+
+  ref = flatpak_installation_get_installed_ref (installation,
+                                                FLATPAK_REF_KIND_APP,
+                                                "com.endlessm.CompanionAppService",
+                                                NULL,
+                                                NULL,
+                                                NULL,
+                                                &error);
+
+  if (ref == NULL)
+    {
+      if (!g_error_matches (error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED))
+        {
+          g_message ("Unexpected error occurred when checking flatpak installation: %s", error->message);
+        }
+
+      ret = FALSE;
+      g_clear_error (&error);
+    }
+
+  g_object_unref (ref);
+  g_object_unref (installation);
+
+  return ret;
+}
+
+
+static void
+cc_sharing_panel_setup_content_sharing_dialog (CcSharingPanel *self)
+{
+  CcSharingPanelPrivate *priv = self->priv;
+  gboolean content_sharing_enabled;
+  GError *error = NULL;
+  GVariant *value;
+  GtkWidget *content_sharing_button = GTK_WIDGET (WID ("content-sharing-button"));
+
+  self->priv->companion_app_avahi_helper_proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SYSTEM,
```

Comment:
As mentioned in my previous review, please move this into `cc_sharing_panel_check_content_sharing_available()` and avoid showing the "Content sharing" list box row altogether when it fails.

Reply:
Hrm, I don't know if I would want to move it into *that* check, `cc_sharing_panel_check_content_sharing_available` is meant to be idempotent, whereas this operation here actually changes the program state by attempting to create the proxy.

In any event, if this fails, the row will not be shown, since the `content_sharing_button` widget is hidden by default:

```
                  <object class="GtkListBoxRow" id="content-sharing-button">
                    <property name="visible">False</property>
```

We only show it if everything succeeds:

```

  /* If everything was successful, show the row */
  gtk_widget_show (content_sharing_button);
  gtk_switch_set_active (GTK_SWITCH (self->priv->content_sharing_switch),
                         content_sharing_enabled);
  g_signal_connect (self->priv->content_sharing_switch, "notify::active",
                    G_CALLBACK (content_sharing_switch_active_changed_cb), self);
```

I think what made this confusing was this leftover:

```
  if (cc_sharing_panel_check_content_sharing_available ())
    cc_sharing_panel_setup_content_sharing_dialog (self);
  else
    gtk_widget_hide (WID ("content-sharing-button"));
```

Which would make the reader think that it was visible by default and needed to be hidden. I'll remove that.